packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -0,0 +1,116 @@
+import * as cdk from '@aws-cdk/core';
+import { TagOption } from '../common';
+import { TagUpdatesOptions } from '../constraints';
+import { IPortfolio } from '../portfolio';
+import { IProduct } from '../product';
+import { CfnPortfolioProductAssociation, CfnResourceUpdateConstraint, CfnTagOption, CfnTagOptionAssociation } from '../servicecatalog.generated';
+import { hashValues } from './util';
+import { InputValidator } from './validation';
+
+export class AssociationManager {
+  public static associateProductWithPortfolio(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct) {
+    const associationKey = hashValues(portfolio.node.addr, product.node.addr, scope.stack.node.addr);
+    if (!this.associationMap.has(associationKey)) {
+      const association = new CfnPortfolioProductAssociation(scope, `PortfolioProductAssociation${associationKey}`, {
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+      });
+
+      const associationValue = new Map<Constraints, CfnConstraintResource>();
+      associationValue.set(Constraints.ASSOCIATION, association);
+      this.associationMap.set(associationKey, associationValue);
+    }
+  }
+
+  public static addResourceUpdateConstraint(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct, options: TagUpdatesOptions) {
+    const associationKey = this.associationPrecheck(scope, portfolio, product);
+    InputValidator.validateLength(this.prettyPrintAssociation(scope, portfolio, product), 'description', 0, 2000, options.description);
+    if (!this.associationMap.get(associationKey)!.has(Constraints.RESOURCE_UPDATE)) {
+      const constraint = new CfnResourceUpdateConstraint(scope, `ResourceUpdateConstraint${associationKey}`, {
+        acceptLanguage: options.acceptedMessageLanguage,
+        description: options.description,
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+        tagUpdateOnProvisionedProduct: options.allowUpdatingProvisionedProductTags === false ? Allowed.NOT_ALLOWED : Allowed.ALLOWED,
+      });
+
+      constraint.addDependsOn(this.associationMap.get(associationKey)!.get(Constraints.ASSOCIATION)!);
+      this.associationMap.get(associationKey)!.set(Constraints.RESOURCE_UPDATE, constraint);
```

Comment:
This expression (`this.associationMap.get(associationKey)!`) is repeated like 3 times - really should avoid doing that.

Reply:
Still repeated 😉.

Reply:
okay, how about this... 
```
  private static checkForConstraint(associationKey: string, constraint: Constraints): boolean {
    if (constraint == Constraints.ASSOCIATION) {
      return !this.associationMap.has(associationKey)
    }
    return !this.associationMap.get(associationKey)!.has(constraint)
  }
```

and then something like 
```
if (this.checkForConstraint(associationKey, Constraints.RESOURCE_UPDATE)) { ...
```

Reply:
The name `check` is too generic. I don't really understand from the name what this does.

Maybe `constraintMissing()`?

Reply:
this is gone now if we use the node child

Reply:
I don't see `node.tryGetChild` used anywhere, and the expression `this.associationMap.get(associationKey)!` is still repeated.

Reply:
the 2nd commit didn't get pushed :/